tools/flang1/flang1exe/semantio.c

```diff
@@ -2895,7 +2895,10 @@ semantio(int rednum, SST *top)
             CNULL);
     }
     if (DTYG(SST_DTYPEG(RHS(1))) == TY_DERIVED &&
-        A_TYPEG(SST_ASTG(RHS(1))) == A_FUNC) {
+        (A_TYPEG(SST_ASTG(RHS(1))) == A_FUNC ||
+        /* Fix bug: if a function reference appears in the array reference. */
```

Comment:
Could you rewrite this comment in a way that explains what the whole conditional block is doing, and why the two different types of checks are needed?

Reply:
> Could you rewrite this comment in a way that explains what the whole conditional block is doing, and why the two different types of checks are needed?

`/* Allocate a temporary ast to store the value of the derived type array
         * reference whose subscript is a function reference, otherwise the
         * function would be incorrectly called in each component I/O. */`
 The comment has been rewritten in the new commit.